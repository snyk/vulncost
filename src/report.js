const sort = ['critical', 'high', 'medium', 'low'];
const sortBy = (a, b) => {
  return sort.indexOf(a.severity) - sort.indexOf(b.severity);
};
const ucFirst = _ => `${_[0].toUpperCase()}${_.slice(1)}`;

/**
 *
 * @param {object} pkg
 * @returns {string} User formatted summary
 */
export const summary = pkg => {
  const message = pkg.vulns.vulnerabilities
    .reduce(
      (acc, curr) => {
        acc[sort.indexOf(curr.severity)]++;
        return acc;
      },
      [0, 0, 0, 0] // relies on the order of critical, high, medium, low (based on `sort`)
    )
    .map((_, i) => `${_} ${sort[i]}`) // remaps to include the severit type
    .filter(_ => _[0] !== '0') // drop those starting with zero
    .join(', ');

  return message;
};

/**
 *
 * @param {string} tested Tested module in format package@version
 * @param {data} data Vulnerability data via test/api
 * @returns {string} User formatted report
 */
export default function report(tested, data) {
  const dedupe = [];

  if (!data.vulnerabilities) {
    return '';
  }

  const res = data.vulnerabilities
    .map(({ severity, upgradePath, title, from, id }) => {
      return {
        id,
        severity,
        upgradePath,
        remediation: upgradePath[0],
        title,
        from: from.pop(),
        direct: upgradePath.length === 1,
      };
    })
    .filter(({ id }) => {
      if (!dedupe.includes(id)) {
        dedupe.push(id);
        return true;
      }
    });

  const direct = res
    .filter(_ => _.direct)
    .sort()
    .sort(sortBy);
  const indirect = res.filter(_ => _.direct === false).sort(sortBy);

  const remediation = res
    .map(_ => _.remediation)
    .sort()
    .pop();

  const lines = ['', `=== ${tested} ===`, ''];

  // NOTE: agreed that UTM doesn't go on visable links (in the output panel)
  if (direct.length) {
    lines.push(
      'Direct:',
      direct
        .map(
          _ =>
            `${ucFirst(_.severity)} ${_.title}\n- https://snyk.io/vuln/${_.id}`
        )
        .join('\n'),
      ''
    );
  }

  if (indirect.length) {
    lines.push(
      'Indirect:',
      indirect
        .map(
          _ =>
            `${ucFirst(_.severity)} ${_.title} in ${
              _.from
            }\n- https://snyk.io/vuln/${_.id}`
        )
        .join('\n'),
      ''
    );
  }

  if (remediation) {
    lines.push('Possible remediation:', remediation);
  } else {
    lines.push('No remediation available.');
  }

  return lines.join('\n');
}

// report('webpack@4.39.0', require('../../test/fixtures/webpack.json')) // ?
