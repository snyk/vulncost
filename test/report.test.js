import report, { summary } from "../src/report"

describe('summary', () => {
  it('should return a well formatted string', () => {
    expect(summary({vulns: {
      vulnerabilities: [
        {
          severity: 'high',
        },
        {
          severity: 'medium',
        },
        {
          severity: 'low',
        }
      ]
    }})).toBe('1 high, 1 medium, 1 low');
  })

  it.each([['critical'], ['high'], ['medium'], ['low']])(
    'supports %s severity',
    severity => {
      expect(
        summary({
          vulns: {
            vulnerabilities: [
              {
                severity,
              },
            ],
          },
        })
      ).toBe(`1 ${severity}`);
    }
  );
})

describe('report', () => {
  it('should return a well formatted report', () => {

    expect(
      report('goof@1.0.1', {
        vulnerabilities: [
          {
            id: 'UNIQ-VULN-ID',
            title: 'my-first-pkg',
            severity: 'low',
            upgradePath: [],
            from: ['loaded-via-pkg'],
          },
        ],
      })
    ).toMatchSnapshot();
  })

  it.each([['critical'], ['high'], ['medium'], ['low']])(
    'supports %s severity',
    severity => {
      expect(
        summary({
          vulns: {
            vulnerabilities: [
              {
                severity,
              },
            ],
          },
        })
      ).toContain(`${severity}`);
    }
  );
})
