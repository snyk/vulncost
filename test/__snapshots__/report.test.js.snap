// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`report should return a well formatted report 1`] = `
"
=== goof@1.0.1 ===

Indirect:
Low my-first-pkg in loaded-via-pkg
- https://snyk.io/vuln/UNIQ-VULN-ID

No remediation available."
`;
